tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+        f.pop();
+        writeln!(f, "}}")?;
+        f.pop();
+        result
+    }
+
+    /// Formats a list of ports as an input or output signature. For example,
+    /// the syntax for ports in the declaration of a component with name-width
+    /// pairs separated by colons is such a signature.
+    ///
+    /// Requires: all ports in `ports` must have definite widths.
+    fn write_ports_sig(
+        &self,
+        f: &mut IndentFormatter<'_, '_>,
+        ports: &[Port],
+    ) -> fmt::Result {
+        for (i, port) in ports.iter().enumerate() {
+            if i > 0 {
+                f.write_str(", ")?;
+            }
+            assert!(!port.has_inferred_width());
+            port.attributes.write(f)?;
+            write!(f, "{}: {}", port.name, port.width)?;
+        }
+        Ok(())
+    }
+}
+
+impl PortProvider for Component {
+    fn get(&self, port: String) -> Port {
+        self.inputs
+            .iter()
+            .chain(self.outputs.iter())
+            .find(|p| p.name == port)
+            .unwrap_or_else(|| {
+                panic!("{}", "port does not exist, violating precondition")
+            })
+            .clone()
+    }
+}
+
+impl AttributeProvider for Component {
+    fn add_attribute(&mut self, attribute: Attribute) {
+        self.attributes.push(attribute);
+    }
+}
+
+impl CalyxWriter for Component {
+    fn marker(&self) -> Option<Marker> {
+        Some(Marker::unique("component", self.name.clone()))
+    }
+
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        self.attributes.write(f)?;
+        write!(
+            f,
+            "{}component {}(",
+            if self.is_comb { "comb " } else { "" },
+            self.name
+        )?;
+        self.write_ports_sig(f, &self.inputs)?;
+        write!(f, ") -> (")?;
+        self.write_ports_sig(f, &self.outputs)?;
+        writeln!(f, ") {{")?;
+        self.brace_section(f, "cells", |f| {
+            for cell in &self.cells {
+                cell.borrow().writeln(f)?;
+            }
+            Ok(())
+        })?;
+        self.brace_section(f, "wires", |f| {
+            for group in &self.groups {
+                group.borrow().writeln(f)?;
+            }
+            for assignment in &self.continuous_assignments {
+                assignment.writeln(f)?;
+            }
+            Ok(())
+        })?;
+        self.brace_section(f, "control", |f| self.control.writeln(f))?;
+        write!(f, "}}")
+    }
+}
+
+/// A complete calyx program containing imports and components. To obtain the
+/// generated calyx as text, use the [`Program::to_string`] function.
+#[derive(Default)]
+pub struct Program {
+    imports: Vec<Import>,
+    /// inv: no element is removed from this array
+    comps: Vec<Rc<RefCell<Component>>>,
+}
+
+impl Program {
+    /// Constructs an empty program.
+    pub fn new() -> Self {
+        Self {
+            imports: vec![],
+            comps: vec![],
+        }
```

Comment:
nit: since you implement `Default` and this is identical to that, you might as
well just call `Default` here